File: 42885.java
import java.util.*;

class Solution {
    public int solution(int[] people, int limit) {
        int answer = 0;
        Arrays.sort(people); // 오름차순 정렬
        int outPerson = 0; // 보트를 타고 나간 사람 수
        
        int minIdx = 0; // 가장 작은
        int maxIdx = people.length -1; // 가장 큰
        
        while (outPerson < people.length) {
            
            if (people[maxIdx] + people[minIdx] <= limit) {
                minIdx ++;
                maxIdx --;
                outPerson += 2;
            } else { 
                // 가장 무거운 사람만 빼기
                maxIdx --;
                outPerson += 1;
            }
            
            answer += 1;
        }
        
        return answer;
    }
}

File: python/42885.py
from collections import deque

def solution(people, limit):
    answer = 0
    people = deque(sorted(people))
    
    while len(people) > 1:
        if people[-1] + people[0] <= limit:
            people.popleft()
            people.pop()
        else:
            people.pop()
        
        answer += 1
    
    if people:
        answer += 1
    
    return answer

#2. 투포인터로 풀어보기
def solution(people, limit):
    people.sort()
    twoPeople = 0 # 두명이 한 번에 건너는 상황
     
    light = 0
    heavy = len(people) - 1
    
    while light < heavy: #투포인터 인덱스는 서로 같아지는 시점 직전에 종료
        if people[light] + people[heavy] <= limit:
            twoPeople += 1
            light += 1
        
        heavy -= 1
        
    return len(people) - twoPeople